### PR TITLE
adapter: move drop cluster code to `catalog_transact`

### PR DESCRIPTION
**Pull request by benesch**
Part of the cluster unification epic (MaterializeInc/cloud#4929).

Follow the style used for dropping other types of objects by making `catalog_transact` responsible for updating the in memory state to reflect the dropped items, rather than the sequencer method.

This will make it easier to turn linked clusters into real clusters, by keeping the logic for dropping clusters and cluster replicas in one place.

### Motivation

   * This PR refactors existing code.

### Tips for reviewer

I wish I could type faster.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a
  companion cloud PR to account for those changes that is tagged with
  the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - n/a

### CODE REVIEW

File: src/adapter/src/catalog.rs [hunk 1]
```diff
@@ -3740,24 +3737,20 @@ impl Catalog {
...
         ops.extend(instance_ops);
-        (ids.into_iter().into_group_map(), ops)
+        ops
     }
 
     /// Creates the catalog operations to drop the given compute instance
     /// replicas.
     ///
     /// Returns the (instance ID, replica ID) pairs to drop, and the catalog
     /// operations to do so.
```
Comment (teskje) [inline src/adapter/src/catalog.rs:3751]: This comment is now outdated.

**Review by teskje — APPROVED** [review]
LGTM!

File: src/adapter/src/catalog.rs (hunk 1 above)
Comment (benesch) [reply src/adapter/src/catalog.rs:3751]: So it is. Fixed!

**Review by aljoscha — APPROVED** [review]
Again mostly following along, but this seems good!

**Comment by jkosh44** [comment]
This all seems sane to me, but the amount of tests failures is a bit scary. I tried to eyeball the code to figure out what's going on and couldn't figure it out.

**Comment by benesch** [comment]
> This all seems sane to me, but the amount of tests failures is a bit scary. I tried to eyeball the code to figure out what's going on and couldn't figure it out.

I was scared too until I figured it out. It turns out to be not scary at all. Linked clusters aren't real until the next PR (https://github.com/MaterializeInc/materialize/pull/17247), so we have to filter out drop optimizations for them. Updated this PR accordingly, and I'm expecting all green.

**Review by jkosh44 — APPROVED** [review]
LGTM

**Pull request merged**